IPlug/CLAP/IPlugCLAP.h

```diff
@@ -0,0 +1,250 @@
+using ClapHost = clap::helpers::HostProxy<clap::helpers::MisbehaviourHandler::Ignore, clap::helpers::CheckingLevel::None>;
+#endif
+
+/** CLAP API base class for an IPlug plug-in
+*   @ingroup APIClasses */
+class IPlugCLAP : public IPlugAPIBase
+                , public IPlugProcessor
+                , public ClapPluginHelper
+{
+  struct ParamToHost
+  {
+    enum class Type { Begin, Value, End };
+
+    ParamToHost()
+      : mType(Type::Value)
+      , mIdx(-1)
+      , mValue(0.0)
+    {}
+
+    ParamToHost(Type type, int idx, double value)
+      : mType(type)
+      , mIdx(idx)
+      , mValue(value)
+    {}
+
+    uint32_t idx() const { return static_cast<uint32_t>(mIdx); }
+    double value() const { return mValue; }
+    uint16_t type() const
+    {
+      switch (mType)
+      {
+        case Type::Begin:   return CLAP_EVENT_PARAM_GESTURE_BEGIN;
+        case Type::End:     return CLAP_EVENT_PARAM_GESTURE_END;
+        case Type::Value:   // fall through
+        default:            return CLAP_EVENT_PARAM_VALUE;
+      }
+    }
+    
+    Type mType;
+    int mIdx;
+    double mValue;
+  };
+  
+public:
+  IPlugCLAP(const InstanceInfo& info, const Config& config);
+
+  // IPlugAPIBase
+  void BeginInformHostOfParamChange(int idx) override;
+  void InformHostOfParamChange(int idx, double normalizedValue) override;
+  void EndInformHostOfParamChange(int idx) override;
+//  void InformHostOfPresetChange() override;
+  bool EditorResize(int viewWidth, int viewHeight) override;
+  
+  // IPlugProcessor
+  void SetTailSize(int tailSize) override;
+  void SetLatency(int samples) override;
+  bool SendMidiMsg(const IMidiMsg& msg) override;
+  bool SendSysEx(const ISysEx& msg) override;
+
+private:
+  
+  // clap_plugin
+  bool init() noexcept override;
+  bool activate(double sampleRate, uint32_t minFrameCount, uint32_t maxFrameCount) noexcept override;
+  void deactivate() noexcept override;
+  bool startProcessing() noexcept override { return true; }
+  void stopProcessing() noexcept override {}
+  clap_process_status process(const clap_process *process) noexcept override;
+  
+  //void onMainThread() noexcept override {}
+
+  // clap_plugin_latency
+  bool implementsLatency() const noexcept override { return true; }
+  uint32_t latencyGet() const noexcept override { return GetLatency(); }
+  
+  // clap_plugin_tail
+  bool implementsTail() const noexcept override { return true; }
+  uint32_t tailGet() const noexcept override;
+
+  // clap_plugin_render
+  bool implementsRender() const noexcept override { return true; }
+  bool renderHasHardRealtimeRequirement() noexcept override { return false; }
+  bool renderSetMode(clap_plugin_render_mode mode) noexcept override;
+  
+  // clap_plugin_state
+  bool implementsState() const noexcept override { return true; }
+  bool stateSave(const clap_ostream *stream) noexcept override;
+  bool stateLoad(const clap_istream *stream) noexcept override;
+  
+  /*void stateMarkDirty() const noexcept {
```

Comment:
@AlexHarker something unfinished here?

Reply:
Iplug2 doesn't have mechanism to mark the state dirty,  but CLAP does - we could add it, but we'd need to discuss what is should be and how it should work.

Reply:
DirtyParametersFromUI() needs to be reworked into DirtyState(). I've learnt recently that the VST3 implementation is problematic, will make an issue